netsync/manager.go

```diff
@@ -1126,30 +1136,76 @@ func (sm *SyncManager) fetchHeaderBlocks() {
 		return
 	}
 
+	queueFetch := func(msg *wire.MsgGetData, quit chan struct{}) {
+		if msg == nil || len(msg.InvList) == 0 {
+			return
+		}
+
+		// Keep fetching until we don't have an error.
```

Comment:
Can we really do this? Is it safe?

Reply:
So `query.WorkManager` is designed around downloading the entire GetData request. This works well for block headers or cfilters but not really for blocks since we verify as they come in.

I wanted to have a guarantee that when you call `queueFetch`, you'll get all the blocks in that request. `queueFetch` will call a new query to `query.WorkManager` which will try to peers to download from.

This seemed to work well in my testing though now I see that you can design to not have this if you make use of `SyncManager.requestedBlocks` map. Though since you'll now need to have a lock around the map to prevent async access.